### PR TITLE
DumpReader - Exposing dump file handles (EnumerateHandles)

### PR DESCRIPTION
**Pull request by Pavel-Durov**
Added MINIDUMP_HANDLES_STREAM class:

MINIDUMP_HANDLES_STREAM uses stream of MINIDUMP_STREAM_TYPE.HandleDataStream type for reading MINIDUMP_HANDLE_DATA_STREAM structure, from which handles information extracted.

GetHandles:
Extracts handles from given stream pointer, returns an IEnumerable of MiniDumps.

MiniDump: wrapped class  for MINIDUMP_HANDLE_DESCRIPTOR, MINIDUMP_HANDLE_DESCRIPTOR_2

### CODE REVIEW

File: src/Microsoft.Diagnostics.Runtime/Utilities/DumpReader.cs [hunk 1]
```diff
@@ -2522,4 +2721,350 @@ internal enum ProcessorArchitecture : ushort
...
+    internal struct MUTEX_ADDITIONAL_INFO_1
+    {
+        public UInt32 Unknown1;
+        public UInt32 Unknown2;
+    };
+
+    internal struct MUTEX_ADDITIONAL_INFO_2
+    {
+        public UInt32 OwnerProcessId;
+
```
Comment (goldshtn) [inline src/Microsoft.Diagnostics.Runtime/Utilities/DumpReader.cs]: Remove this blank line.

File: src/Microsoft.Diagnostics.Runtime/Utilities/DumpReader.cs [hunk 2]
```diff
@@ -2522,4 +2721,350 @@ internal enum ProcessorArchitecture : ushort
...
+        public UInt32 Unknown4;
+        public UInt32 Unknown5;
+        public UInt32 ProcessId;
+        public UInt32 ParentProcessId;
+        public UInt32 Unknown6;
+    };
+
+#else
+
+    struct MUTEX_ADDITIONAL_INFO_2
```
Comment (goldshtn) [inline src/Microsoft.Diagnostics.Runtime/Utilities/DumpReader.cs]: The previous structs are marked explicitly as internal, and these ones aren't. Let's be consistent.

File: src/Microsoft.Diagnostics.Runtime/Utilities/DumpReader.cs [hunk 3]
```diff
@@ -382,6 +384,186 @@ public static bool MiniDumpReadDumpStream(IntPtr pBase, MINIDUMP_STREAM_TYPE typ
...
+                                handle.OwnerThreadId = mutexInfoM2.OwnerThreadId;
+                            }
+                            break;
+                        case MINIDUMP_HANDLE_OBJECT_INFORMATION_TYPE.MiniProcessInformation1:
+                            handle.Type = DumpHandleType.PROCESS1;
+                            break;
+                        case MINIDUMP_HANDLE_OBJECT_INFORMATION_TYPE.MiniProcessInformation2:
+                            {
+                                handle.Type = DumpHandleType.PROCESS2;
+                                var processInfro2 = dumpPointer.PtrToStructure<PROCESS_ADDITIONAL_INFO_2>();
```
Comment (goldshtn) [inline src/Microsoft.Diagnostics.Runtime/Utilities/DumpReader.cs]: Typo in variable name.

File: src/Microsoft.Diagnostics.Runtime/Utilities/DumpReader.cs [hunk 4]
```diff
@@ -382,6 +384,186 @@ public static bool MiniDumpReadDumpStream(IntPtr pBase, MINIDUMP_STREAM_TYPE typ
...
+                    switch (pObjectInfo.InfoType)
+                    {
+                        case MINIDUMP_HANDLE_OBJECT_INFORMATION_TYPE.MiniHandleObjectInformationNone:
+                            handle.Type = DumpHandleType.NONE;
+                            break;
+                        case MINIDUMP_HANDLE_OBJECT_INFORMATION_TYPE.MiniThreadInformation1:
+                            {
+                                handle.Type = DumpHandleType.THREAD;
+
+                                var threaInfo = dumpPointer.PtrToStructure<THREAD_ADDITIONAL_INFO>();
```
Comment (goldshtn) [inline src/Microsoft.Diagnostics.Runtime/Utilities/DumpReader.cs]: Typo in variable name.

File: src/Microsoft.Diagnostics.Runtime/Utilities/DumpReader.cs [hunk 5]
```diff
@@ -382,6 +384,186 @@ public static bool MiniDumpReadDumpStream(IntPtr pBase, MINIDUMP_STREAM_TYPE typ
...
+            public class MINIDUMP_HANDLES_STREAM
+            {
+                public MINIDUMP_HANDLES_STREAM(DumpPointer pStream, DumpReader contex)
+                {
+                    _pStream = pStream;
+                    _contex = contex;
+                }
+
+                DumpPointer _pStream;
+                DumpReader _contex;
```
Comment (goldshtn) [inline src/Microsoft.Diagnostics.Runtime/Utilities/DumpReader.cs]: I assume you meant `_context`?

File: src/Microsoft.Diagnostics.Runtime/Utilities/DumpReader.cs [hunk 6]
```diff
@@ -382,6 +384,186 @@ public static bool MiniDumpReadDumpStream(IntPtr pBase, MINIDUMP_STREAM_TYPE typ
...
+                /// <param name="pointer">DumpPointer to stream</param>
+                /// <returns></returns>
+                private DumpHandle GetHandleData(MINIDUMP_HANDLE_DESCRIPTOR_2 handle, DumpPointer pointer)
+                {
+                    string objectName, typeName;
+                    typeName = objectName = null;
+
+                    if (handle.ObjectNameRva != 0)
+                    {
+                        objectName = _contex.GetString(new DumpNative.RVA() { Value = (uint)handle.ObjectNameRva });
```
Comment (goldshtn) [inline src/Microsoft.Diagnostics.Runtime/Utilities/DumpReader.cs]: Why the cast to `uint`? Isn't it possible for the RVA to be greater than 4GB in large dumps?

File: src/Microsoft.Diagnostics.Runtime/Utilities/DumpReader.cs [hunk 7]
```diff
@@ -2522,4 +2721,350 @@ internal enum ProcessorArchitecture : ushort
...
+        /// <summary>
+        /// Second mutex unknown field
+        /// </summary>
+        public object Field2 { get; internal set; }
+    }
+
+    /// <summary>
+    /// Handle additional object info constructed by MINIDUMP_HANDLE_OBJECT_INFORMATION structure 
+    /// </summary>
+    public class DumpHandleInfo
```
Comment (goldshtn) [inline src/Microsoft.Diagnostics.Runtime/Utilities/DumpReader.cs:2911]: How are clients expected to use this? Are they supposed to look at `InfoType` and then decide which of the following internal structures to read? If so, there should be a pointer somewhere in this class pointing to where that structure is in the dump.

**Comment by goldshtn** [comment]
Looks good! I've made a few comments on the diff, please address them and we'll be good to merge.

**Comment by Pavel-Durov** [comment]
Casting ObjectNameRva to uint:
objectName = _contex.GetString(new DumpNative.RVA() { Value = (uint)handle.ObjectNameRva });
I am casting it to uint only because the GetString function accepts DumpNative.RVA type parameter with Value property of type uint.

DumpHandleInfo class:
DumpHandleInfo is just a wrapper class to MINIDUMP_HANDLE_OBJECT_INFORMATION structure, I really don't know how the clients should use it, the only information that I get from MINIDUMP_HANDLE_OBJECT_INFORMATION InfoType, NextInfoRva and sizeOfInfo, no pointers to actual structures. I am obtaining this information by do-while iteration based on NextInfoRva offset.

**Comment by goldshtn** [comment]
OK, so do you think the RVA type has to be `ulong`? I modified `DumpReader` because it didn't support >4GB dumps at all, perhaps I should have changed that part as well. If RVA is an address within the dump, then yeah -- it can be >4GB, and has to be a `ulong`.

Re the `DumpHandleInfo` class, you actively read information from it based on its type and assign fields like the mutex owner, process id, thread id, that kind of stuff. It's just that the exact variant of the structure you're getting depends on the type field -- which is fine, and you can have users extract it themselves, but you need to expose the address of the struct from which they can extract it. It's the address from which you're getting the [`MINIDUMP_HANDLE_OBJECT_INFORMATION`](https://msdn.microsoft.com/en-us/library/windows/desktop/ms680375%28v=vs.85%29.aspx) structure.

**Pull request merged**